cpp/DEVELOPER_GUIDE.md

```diff
@@ -5,9 +5,22 @@ This document summarizes rules and best practices for contributions to the cuML
 Please start by reading [CONTRIBUTING.md](../CONTRIBUTING.md).
 
 ## Thread safety
-cuML is thread safe so its functions can be called from multiple host threads if they use different handles.
 
-The implementation of cuML is single threaded.
+With the exception of the cumlHandle, cuML algorithms should maintain thread-safety and are, in general, 
+assumed to be single threaded. This means they should be able to be called from multiple host threads so 
+long as different instances of `cumlHandle` are used.
+
+Exceptions are made for algorithms that can take advantage of multiple CUDA streams in order to oversubscribe 
+or increase occupancy on a single GPU. In these cases, the use of multiple host threads within cuML algorithms
+should be used only to maintain concurrency of the underlying CUDA streams. Multiple host threads should be 
+used sparingly, be bounded, and should steer clear of performing CPU-intensive computations.
+
+A good example of an acceptable use of host threads can be found in [this blog article](https://devblogs.nvidia.com/gpu-pro-tip-cuda-7-streams-simplify-concurrency/)
```

Comment:
By linking that blog post are you suggesting that cuML should or does use CUDA's per-thread default stream?

Reply:
@harrism, there were two reasons I linked that blog post, actually. Knowing that we can use the default stream per thread is useful and it might be worth adding it as an option in our build. The other reason I linked to it is that it provides an example of this more general design:

```
cudaStreamSynchronize(streamInCurrentHostThread);

#pragma omp parallel for num_threads(n_threads)
for(int i = 0; i < n; i++) {
    ... possible light cpu pre-processing ...
    cudaStream_t s;
    cudaStreamCreate(&s);
    my_kernel1<<<b, tpb, 0, s>>>(...);
    ...
    ... some possible async d2h / h2d copies ...
    my_kernel2<<<b, tpb, 0, s>>>(...);
    ...
    cudaStreamSynchronize(s);
    ... possible light cpu post-processing ...
}
```

I have not seen an abundance of formal examples for this design. Maybe I should also include this code snippet?